test/schema-feature.test.js

```diff
@@ -0,0 +1,445 @@
+    z: { $id: 'z', my: 'schema' },
+    a: { $id: 'a', my: 'schema' },
+    b: { $id: 'b', my: 'schema' },
+    c: { $id: 'c', my: 'schema', properties: { a: 'a', b: 1 } }
+  }
+
+  fastify.addSchema(schemas.z)
+
+  fastify.register((instance, opts, next) => {
+    instance.addSchema(schemas.a)
+    pluginDeepOneSide = instance
+    next()
+  })
+
+  fastify.register((instance, opts, next) => {
+    instance.addSchema(schemas.b)
+    instance.register((subinstance, opts, next) => {
+      subinstance.addSchema(schemas.c)
+      pluginDeepTwo = subinstance
+      next()
+    })
+    pluginDeepOne = instance
+    next()
+  })
+
+  fastify.ready(err => {
+    t.error(err)
+    t.deepEqual(fastify.getSchemas(), [schemas.z])
+    t.deepEqual(pluginDeepOneSide.getSchemas(), [schemas.z, schemas.a])
+    t.deepEqual(pluginDeepOne.getSchemas(), [schemas.z, schemas.b])
+    t.deepEqual(pluginDeepTwo.getSchemas(), [schemas.z, schemas.b, schemas.c])
+  })
+})
+
+test('Use the same schema id in different places', t => {
+  t.plan(1)
+  const fastify = Fastify()
+
+  fastify.addSchema({
+    $id: 'test',
+    type: 'object',
+    properties: {
+      id: { type: 'number' }
+    }
+  })
+
+  fastify.get('/:id', {
+    handler: echoParams,
+    schema: {
+      response: {
+        200: {
+          type: 'array',
+          items: { $ref: 'test#/properties/id' }
+        }
+      }
+    }
+  })
+
+  fastify.post('/:id', {
+    handler: echoBody,
+    schema: {
+      body: { id: { $ref: 'test#/properties/id' } },
+      response: {
+        200: { id: { $ref: 'test#/properties/id' } }
+      }
+    }
+  })
+
+  fastify.ready(err => t.error(err))
+})
+
+test('Get schema anyway should not add `properties` if allOf is present', t => {
+  t.plan(1)
+  const fastify = Fastify()
+
+  fastify.addSchema({
+    $id: 'first',
+    type: 'object',
+    properties: {
+      first: { type: 'number' }
+    }
+  })
+
+  fastify.addSchema({
+    $id: 'second',
+    type: 'object',
+    allOf: [
+      {
+        type: 'object',
+        properties: {
+          second: { type: 'number' }
+        }
+      },
+      fastify.getSchema('first')
+    ]
+  })
+
+  fastify.get('/', {
+    handler: () => {},
+    schema: {
+      querystring: fastify.getSchema('second'),
+      response: { 200: fastify.getSchema('second') }
+    }
+  })
+
+  fastify.ready(err => t.error(err))
+})
+
+test('Get schema anyway should not add `properties` if oneOf is present', t => {
+  t.plan(1)
+  const fastify = Fastify()
+
+  fastify.addSchema({
+    $id: 'first',
+    type: 'object',
+    properties: {
+      first: { type: 'number' }
+    }
+  })
+
+  fastify.addSchema({
+    $id: 'second',
+    type: 'object',
+    oneOf: [
+      {
+        type: 'object',
+        properties: {
+          second: { type: 'number' }
+        }
+      },
+      fastify.getSchema('first')
+    ]
+  })
+
+  fastify.get('/', {
+    handler: () => {},
+    schema: {
+      querystring: fastify.getSchema('second'),
```

Comment:
@mcollina writing test, I thought that this feature `.getSchema` could be a trade-off between removing the shared schema and keep a utility to archive the same results (but you will need the `fastify` instance instead of a plain string)